cmd/src/search_stream.go

```diff
@@ -21,20 +22,111 @@ func init() {
 }
 
 func streamSearch(query string, opts streaming.Opts, client api.Client, w io.Writer) error {
-	t, err := parseTemplate(streamingTemplate)
-	if err != nil {
-		panic(err)
+	var d streaming.Decoder
+	if opts.Json {
+		d = jsonDecoder(w)
+	} else {
+		t, err := parseTemplate(streamingTemplate)
+		if err != nil {
+			return err
+		}
+		d = textDecoder(query, t, w)
+	}
+	return streaming.Search(query, opts, client, d)
+}
+
+// jsonDecoder streams results as JSON to w.
+func jsonDecoder(w io.Writer) streaming.Decoder {
+	// check is used whenever we write to w. If a write fails, the output is most
+	// likely not a valid JSON and we should stop the stream immediately. Since the
+	// handlers of the decoder don't handle errors we use panic.
+	check := func(err error) {
+		if err != nil {
+			panic(err)
+		}
+	}
+	_, err := w.Write([]byte("{\"Results\":[\n"))
+	check(err)
+
+	first := true
+	p := streaming.Progress{}
+	a := make([]*streaming.EventAlert, 0)
+
+	// write writes a key-value pair k, v to w; write will return an error if v
+	// cannot be JSON-encoded. The caller is responsible for adding a trailing comma
+	// if necessary.
+	write := func(k string, v interface{}) error {
+		b, err := marshalIndent(v)
+		if err != nil {
+			return err
+		}
+		_, err = w.Write([]byte(fmt.Sprintf("\"%s\":", k)))
+		if err != nil {
+			return err
+		}
+		_, err = w.Write(b)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+
+	comma := func() error {
+		_, err := w.Write([]byte(",\n"))
+		return err
 	}
-	logError := func(msg string) {
-		_, _ = fmt.Fprintf(os.Stderr, msg)
+
+	return streaming.Decoder{
+		OnProgress: func(progress *streaming.Progress) {
+			if !progress.Done {
+				return
+			}
+			p = *progress
+		},
+		OnMatches: func(matches []streaming.EventMatch) {
+			if first {
+				first = false
+			} else {
+				check(comma())
+			}
+			b, err := marshalIndent(matches)
+			check(err)
+
+			// With each event we append to the list of results, hence we cut delimiters here
+			// and add them back later in OnDone.
+			b = bytes.Trim(b, "[]\n")
+			_, err = w.Write(b)
+			check(err)
+		},
+		OnAlert: func(alert *streaming.EventAlert) {
+			a = append(a, alert)
+		},
+		OnError: func(eventError *streaming.EventError) {
+			logError(eventError.Message)
+		},
+		OnDone: func() {
```

Comment:
as I mentioned in the other PR, why don't you just do this after `streaming.Decoder` returns, rather than creating a `streaming.Decoder` that writes out json. IE I would suspect the interface here is you take in a `io.Reader` and return an `io.Writer`. The `io.Reader` is the search response body, the `io.Writer` is os.Stderr (or piped to it) and will write out json?

Reply:
> why don't you just do this after streaming.Decoder returns
Do you mean when ReadAll returns? 

IMO the "done" event is an event like any other and we should be able to hook into it. Like for example in this PR, I call `ReadAll` in a totally different place and it would be strange closing the JSON object there. 

Sure, if we decide not to stream out 1 object but 1 object per result then we don't need the hook.